src/lib/x509/x509_ext.cpp

```diff
@@ -868,6 +877,607 @@ const std::string& TNAuthList::Entry::telephone_number() const {
+   }
+   BER_Decoder obj_ber = BER_Decoder(obj);
+
+   std::vector<uint8_t> addr_family;
+   obj_ber.decode(addr_family, ASN1_Type::OctetString);
+   const size_t addr_family_length = addr_family.size();
+
+   if(addr_family_length != 2 && addr_family_length != 3) {
+      throw Decoding_Error("(S)AFI can only contain 2 or 3 bytes");
+   }
+
+   m_afi = (addr_family[0] << 8) | addr_family[1];
+
+   if(addr_family_length == 3) {
+      m_safi = addr_family[2];
+   }
+
+   if(m_afi == 1) {
+      IPAddressChoice<Version::IPv4> addr_choice;
+      obj_ber.decode(addr_choice);
+      m_ip_addr_choice = addr_choice;
+   } else if(m_afi == 2) {
+      IPAddressChoice<Version::IPv6> addr_choice;
+      obj_ber.decode(addr_choice);
+      m_ip_addr_choice = addr_choice;
+   } else {
+      throw Decoding_Error("Only AFI IPv4 and IPv6 are supported.");
+   }
+}
+
+void IPAddressBlocks::sort_and_merge() {
+   // Sort IPAddressFamilies by afi/safi values
+   //
+   // see: https://www.rfc-editor.org/rfc/rfc3779.html#section-2.2.3.3
+   //
+   // families with no safis are ordered before families with safis,
+   // v4 families are ordered before v6 families (i.e. they are sorted by afis)
+   //
+   // families with the same afi/safi combination are then merged
+
+   // std::map is ordered, so a uint32_t of [afi,safi] will be in the right order
+   std::map<uint32_t, std::vector<IPAddressFamily>> afam_map;
+   for(IPAddressFamily& block : m_ip_addr_blocks) {
+      uint32_t afam = block.afi();
+      if(block.safi().has_value()) {
+         afam = static_cast<uint32_t>(afam << 8) | block.safi().value();
+      }
+      std::vector<IPAddressFamily>& fams = afam_map[afam];
+      fams.push_back(block);
+   }
+
+   std::vector<IPAddressFamily> merged_blocks;
+   for(auto it = afam_map.begin(); it != afam_map.end(); it++) {
+      // fams consists of families with the same afi/safi combination
+      std::vector<IPAddressFamily>& fams = it->second;
+      // since at least 1 block has to belong to a afi/safi combination for it to appear in the map,
+      // fams cannot be empty
+      BOTAN_ASSERT_NOMSG(!fams.empty());
+
+      // fams[0] has to have the same choice type as the fams in the same bucket
+      if(std::holds_alternative<IPAddressChoice<Version::IPv4>>(fams[0].addr_choice())) {
+         merged_blocks.push_back(merge<Version::IPv4>(fams));
+      } else {
+         merged_blocks.push_back(merge<Version::IPv6>(fams));
+      }
+   }
+   m_ip_addr_blocks = merged_blocks;
+}
+
+template <IPAddressBlocks::Version V>
+IPAddressBlocks::IPAddressFamily IPAddressBlocks::merge(std::vector<IPAddressFamily>& blocks) {
+   // Merge IPAddressFamilies that have the same afi/safi combination
+   //
+   // see: https://www.rfc-editor.org/rfc/rfc3779.html#section-2.2.3.3
+
+   if(blocks.empty()) {
+      throw Exception("at least 1 block is needed to merge");
+   }
```

Comment:
If this should not occur absent a bug, this should be a `BOTAN_ASSERT`. If there is some circumstance where this is expected to occur, the error message should be clearer.